src/annotator.cpp

```diff
@@ -0,0 +1,361 @@
+        if (!id.empty()) {
+            idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::RESET_VALUE, component->reset(r))));
+            // idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::RESET_VALUE, component->reset(r)->resetValue())));
+        }
+    }
+
+    for (size_t c = 0; c < component->componentCount(); ++c) {
+        listComponentIdsAndItems(component->component(c), idList);
+    }
+}
+
+ItemList listIdsAndItems(const ModelPtr &model)
+{
+    // Collect all existing ids in a list and return. NB can't use a map or a set as we need to be able to print
+    // invalid models (with duplicated ids) too.
+
+    ItemList idList;
+    // Model.
+    std::string id = model->id();
+    if (!id.empty()) {
+        idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::MODEL, model)));
+    }
+    // Units.
+    for (size_t u = 0; u < model->unitsCount(); ++u) {
+        id = model->units(u)->id();
+        if (!id.empty()) {
+            idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::UNITS, model->units(u))));
+        }
+        for (size_t i = 0; i < model->units(u)->unitCount(); ++i) {
+            std::string prefix;
+            std::string reference;
+            double exponent;
+            double multiplier;
+            model->units(u)->unitAttributes(i, reference, prefix, exponent, multiplier, id);
+            if (!id.empty()) {
+                idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::UNIT, std::make_pair(model->units(u), i))));
+            }
+        }
+        if (model->units(u)->isImport() && model->units(u)->importSource() != nullptr) {
+            id = model->units(u)->importSource()->id();
+            idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::IMPORT, model->units(u)->importSource())));
+        }
+    }
+    // Components.
+    for (size_t c = 0; c < model->componentCount(); ++c) {
+        listComponentIdsAndItems(model->component(c), idList);
+    }
+    // Encapsulation.
+    id = model->encapsulationId();
+    if (!id.empty()) {
+        idList.insert(std::make_pair(id, std::make_pair(Annotator::Type::ENCAPSULATION, std::string())));
+    }
+
+    return idList;
+}
+
+void Annotator::build(const ModelPtr &model)
+{
+    mPimpl->isBuilt = false;
+    removeAllIssues();
+    mPimpl->mIdList.clear();
+    mPimpl->mIdList = listIdsAndItems(model);
+    mPimpl->isBuilt = true;
+    mPimpl->mModel = model;
+}
+
+AnyItem Annotator::item(const std::string &id)
+{
+    AnyItem item;
+    if (!mPimpl->isBuilt) {
+        auto issue = libcellml::Issue::create();
+        issue->setDescription("Please call the Annotator::build function before attempting to access items by their id.");
+        issue->setLevel(libcellml::Issue::Level::ERROR);
+        addIssue(issue);
+        item = std::make_pair(Annotator::Type::ISSUE, issue);
+        return item;
+    }
+    if (mPimpl->mIdList.empty()) {
```

Comment:
Not sure this check is needed. If you can't find the requested `id`, it doesn't really matter whether there are no IDs or not. At the very least, I don't see how this is an error, if anything a hint or maybe a warning...but would just remove this block.

Reply:
It needs to be there to distinguish between situations where the annotator has been built (the previous check) and where zero items with ids exist in the model.  Otherwise, we can't tell the difference between an empty idList.  It doesn't need to be an error, but it does need to be a return path.

Reply:
83ce6a11

Reply:
but why do we need to distinguish the case when there are no items with `id`s the model vs the requested `id` is not used? There is absolutely nothing wrong with that, so I'm not sure why we need to care?

This use of `any` is really looking less and less attractive, as it seems to be forcing us to do things that just don't make sense...

Reply:
There's not a lot of difference, other than telling the user that there's no point looking for any ids, as there are none, rather than just not having the single one it's looking for.  I can lump them together if it's clearer.

Reply:
83ce6a11